api/server.go

```diff
@@ -55,16 +62,26 @@ func NewServer(APIaddr string, requiredUserAgent string, cs modules.ConsensusSet
 	return srv, nil
 }
 
-// Serve listens for and handles API calls. It a blocking function.
+// Serve listens for and handles API calls. It is a blocking function.
 func (srv *Server) Serve() error {
+	// Block the Close() method until Serve() has finished.
+	srv.wg.Add(1)
+	defer srv.wg.Done()
+
 	// stop the server if a kill signal is caught
-	sigChan := make(chan os.Signal)
+	sigChan := make(chan os.Signal, 1)
 	signal.Notify(sigChan, os.Interrupt, os.Kill)
-	defer signal.Reset(os.Interrupt, os.Kill)
+	defer signal.Stop(sigChan)
+	stop := make(chan struct{})
+	defer close(stop)
```

Comment:
For reference, [my old shutdown proposal](https://github.com/NebulousLabs/Sia/pull/898#issuecomment-170403863).